### PR TITLE
Promote tools/compose to xkbcli-compile-compose

### LINKED ISSUE #381: Make tools/compose part of the xkbcli commands
This is a follow-up of #367.

Convert `tools/compose.c` to `xkbcli compose`:

- Better names for the `--locale` arguments.
- ~Add `--format` (currently only text-v1).~ add it only when there is a new format?
- Check return value of `printfs`.
- Add to `xkbcli`.
- Test it in `test/tool-option-parsing.py`.
- Write `man` page

### PR DESCRIPTION
**Pull request by wismill**
Fixes #381

### CODE REVIEW

File: tools/compile-compose.c [hunk 1]
```diff
@@ -69,10 +147,17 @@ print_compose_table_entry(struct xkb_compose_table_entry *entry)
...
     if (*utf8 != '\0') {
-        printf(" \"%s\"", utf8);
+        if (!escape_utf8_string(utf8, buf, sizeof(buf))) {
+            fprintf(stderr,
+                    "ERROR: Cannot escape the string: "
+                    "unsufficient buffer size.\n");
+            printf("\"\"");
+        } else {
+            printf(" \"%s\"", buf);
+        }
```
Comment (whot) [inline tools/compile-compose.c:156]: Re-using `buf` here feels naughty. Any reason we can't just return an allocated buffer from `escape_utf8_string()` and worry about greater things in life than bufsizes? :)


If you alloc a buffer `4 * strlen(from) + 1` then our string will definitely fit and we can realloc to the smaller size before return.

File: tools/xkbcli-bash-completion.sh [hunk 2]
```diff
@@ -95,6 +95,8 @@ ___xkbcli_subcommand()
     # • https://github.com/scop/bash-completion/blob/fdf4456186eb4548ef628e65fb1be73d8e4695e9/bash_completion.d/000_bash_completion_compat.bash#L311
     # We need both as the current help messages adopt both GNU and BSD styles.
     opts=$(_parse_usage xkbcli "$1 --help")
+    opts+="
+    "
```
Comment (whot) [inline tools/xkbcli-bash-completion.sh:99]: Can you add a comment to the commit message to explain what it fixes? Some error message maybe?

File: meson.build [hunk 3]
```diff
@@ -460,6 +460,7 @@ if build_tools
                                        install: true,
                                        install_dir: dir_libexec)
     install_man('tools/xkbcli-compile-keymap.1')
+    configh_data.set10('HAVE_XKBCLI_COMPILE_KEYMAP', true)
```
Comment (whot) [inline meson.build:463]: This needs to be a separate commit, it seems unrelated to the compose tool here.

**Review by whot — APPROVED** [review]
nitpicks, overall LGTM, thanks. @bluetech may have a comment on why the locale handling was implemented like that first though.

File: tools/compile-compose.c [hunk 4]
```diff
@@ -56,6 +59,81 @@ usage(FILE *fp, char *progname)
...
+            f += 4;
+            previous_hex_escape = false;
+        }
+    }
+    if (t >= size)
+        return false;
+    to[t] = '\0';
+    return true;
+}
+
```
[the remark this answers is not in the document]
Comment (wismill) [reply tools/compile-compose.c:132]: Done

[the remark this answers is not in the document]
Comment (wismill) [reply tools/compile-compose.c:132]: Done. I added a test for random strings plus a few static strings.

**Review by wismill — COMMENTED** [review]
@whot thanks for the review. I fixed the code, but I am not sure where to put the escape utility function.

**Comment by whot** [comment]
> I am not sure where to put the escape utility function.

based on the filename `src/utf8.c` seems like a good match? :)

**Comment by wismill** [comment]
I realized I did not push my changes…

> based on the filename `src/utf8.c` seems like a good match? :)

@whot I would prefer to keep `src/utf8.c` only related to low-level encoding. So I chose `src/compose/dump.h`. Note that it could be adapted to be used in other contexts: it would need a flag to indicate whether to use hexadecimal escapes  (compose-only) or octal ones.

File: src/compose/dump.h [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+#ifndef COMPOSE_DUMP_H
+#define COMPOSE_DUMP_H
+
+#include "config.h"
+
+#include <stdlib.h>
+
+#include "src/utils.h"
+
+static char*
```
Comment (whot) [inline src/compose/dump.h:44]:
```suggestion
static inline char*
```

Not needed right now but it's a common workaround to stop compiler warnings about unused statics if this header file is included somewhere but not all functions in it are used.

File: src/compose/dump.h [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            nbytes = 4;
+        memcpy(&to[t], &from[f], nbytes);
+        t += nbytes;
+        f += nbytes;
+        previous_hex_escape = false;
+    }
+    to[t++] = '\0';
+    if (t < max_new_length)
+        return realloc(to, t);
+    return to;
```
Comment (whot) [inline src/compose/dump.h:95]:
```suggestion
   return realloc(to, t);
```
simpler and for 99.9% of all real-world strings we realloc anyway.

File: test/compose.c [hunk 7]
```diff
@@ -769,20 +772,111 @@ test_traverse(struct xkb_context *ctx)
...
+
+    /* Test round-trip of random strings */
+#   define SAMPLE_SIZE 1000
+#   define MIN_CODE_POINT 0x0001
+#   define MAX_CODE_POINTS_COUNT 15
+    char buf[1 + MAX_CODE_POINTS_COUNT * 4];
+    for (int ascii = 1; ascii >= 0; ascii--) {
+        for (size_t s = 0; s < SAMPLE_SIZE; s++) {
+            /* Create the string */
+            size_t length = 1 + (rand() % MAX_CODE_POINTS_COUNT);
```
Comment (whot) [inline test/compose.c:827]: if we're using `rand()` in tests, best to print the seed to the logs so it can be reproduced - otherwise you just get an error with no idea how to get to that same point again. So something like this in `main`

```c
int seed = time();
printf("Using random seed: %d\n", seed);
srand(seed);
```

**Review by whot — COMMENTED** [review]
sorry, got no focus today so only two comments but I trust you enough that this will be fine.

File: test/compose.c [hunk 8]
```diff
@@ -769,20 +772,111 @@ test_traverse(struct xkb_context *ctx)
...
+                /* Handle invalid code point in UTF-8 */
+                if (!count)
+                    /* Use replacement character: � */
+                    count = utf32_to_utf8(0xfffd, &buf[c]);
+                c += count - 1;
+                assert(c <= sizeof(buf) - 1);
+            }
+            assert(buf[c] == '\0');
+            assert(strlen(buf) == c);
+            assert(is_valid_utf8(buf, c));
```
Comment (whot) [inline test/compose.c:841]: unrelated to this series: we should have something like
```c
#define assert_printf(cond, ...) \
   if (!(cond)) { \
      printf("Assertion failure: __VA_ARGS__); \
      assert(cond); \
   }
```
to aid with debugging assertion failures.

**Review by wismill — COMMENTED** [review]
Fixed reviews comments and added a few comments.

@bluetech could you check my refactoring of locale handling?

File: tools/compile-compose.c [hunk 9]
```diff
@@ -48,12 +48,8 @@ usage(FILE *fp, char *progname)
             " --file FILE\n"
             "    Specify a Compose file to load\n"
             " --locale LOCALE\n"
-            "    Specify the locale directly\n"
-            " --locale-from-env\n"
```
Comment (bluetech) [inline tools/compile-compose.c]: The reason for adding this was mostly for testing vs. setlocale, but I agree it doesn't serve a purpose anymore.

**Review by bluetech — APPROVED** [review]
Nice, it would be nice to have `xkbcli compile-compose`.

I left a few minor comments, but LGTM (feel free to merge after fixing).

File: src/compose/dump.h [hunk 10]
```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright © 2013 Pierre Le Marre <dev@wismill.eu>
```
[the remark this answers is not in the document]
Comment (wismill) [reply src/compose/dump.h]: 😅

File: src/compose/dump.h [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+ * It also escapes hexadecimal digits after an hexadecimal escape. This is not
+ * strictly needed by the current implementation: "\x0abcg" parses as "␊bcg",
+ * but better be cautious than sorry and produce "\x0a\x62\x63g" instead.
+ * In the latter string there is no ambiguity and no need to know the maximum
+ * number of digits supported by the escape sequence.
+ */
+static inline char*
+escape_utf8_string_literal(const char *from)
+{
+    const int length = strlen(from);
```
[the remark this answers is not in the document]
Comment (wismill) [reply src/compose/dump.h]: 👍

File: src/compose/parser.c [hunk 12]
```diff
@@ -468,6 +468,22 @@ resolve_modifier(const char *name)
     return XKB_MOD_INVALID;
 }
 
+/* For testing purposes */
```
[the remark this answers is not in the document]
Comment (wismill) [reply src/compose/parser.c]: Good idea

File: tools/compile-compose.c [hunk 13]
```diff
@@ -69,10 +69,17 @@ print_compose_table_entry(struct xkb_compose_table_entry *entry)
...
         }
     }
-    printf(":");
+    printf(" : ");
     const char *utf8 = xkb_compose_table_entry_utf8(entry);
     if (*utf8 != '\0') {
-        printf(" \"%s\"", utf8);
+        char *escaped = escape_utf8_string_literal(utf8);
+        if (!escaped) {
+            fprintf(stderr, "ERROR: Cannot escape the string: allocation error\n");
```
[the remark this answers is not in the document]
Comment (wismill) [reply tools/compile-compose.c:73]: Good point! And I realized that the current exit code is always… non-zero! Fixed.

File: tools/compile-compose.c (hunk 9 above)
Comment (wismill) [reply tools/compile-compose.c]: Duly noted. I modified the commit message to explain that.
